File: cpp/cpp_100-999/654.cc
// C++
// leetcode 654
// https://leetcode.cn/problems/maximum-binary-tree/

/**
 * Definition for a binary tree node.
 * struct TreeNode {
 *     int val;
 *     TreeNode *left;
 *     TreeNode *right;
 *     TreeNode() : val(0), left(nullptr), right(nullptr) {}
 *     TreeNode(int x) : val(x), left(nullptr), right(nullptr) {}
 *     TreeNode(int x, TreeNode *left, TreeNode *right) : val(x), left(left), right(right) {}
 * };
 */
class Solution {
public:
    TreeNode* constructMaximumBinaryTree(vector<int>& nums) {
    	return helper(nums, 0, nums.size() - 1);
    }

    TreeNode* helper(vector<int>& nums, int l, int r) {
    	if (l > r) {
    		return nullptr;
    	}

    	int mx = INT_MIN, p = -1;
    	for (int i = l; i <= r; i++) {
    		if (nums[i] > mx) {
    			mx = nums[i];
    			p = i;
    		}
    	}

    	TreeNode *res = new TreeNode(mx);
    	res -> left = helper(nums, l, p -1); 
    	res -> right = helper(nums, p + 1, r);
    	return res; 
    }
};

File: Go/Go_100-999/654.go
// Go
// leetcode 654
// https://leetcode.cn/problems/maximum-binary-tree/

/**
 * Definition for a binary tree node.
 * type TreeNode struct {
 *     Val int
 *     Left *TreeNode
 *     Right *TreeNode
 * }
 */
func constructMaximumBinaryTree(nums []int) *TreeNode {
	var hepler func(arr []int, l, r int) *TreeNode
	hepler = func(arr []int, l, r int) *TreeNode {
		// 返回空指针
		if l > r {
			return nil
		}
		// 找到 [l, r] 范围内最大值
		mx, p := -1, -1
		for i := l; i <= r; i++ {
			if arr[i] > mx {
				mx = arr[i]
				p = i
			}
		}

		return &TreeNode{mx, hepler(arr, l, p-1), hepler(arr, p+1, r)}
	}

	return hepler(nums, 0, len(nums)-1)
}